src/components/implementation/torrent/ramfs/ramfs.c

```diff
@@ -186,6 +278,160 @@ twrite(spdid_t spdid, td_t td, int cbid, int sz)
+
+	fso = t->data;
+	// we aren't tracking these right now. Need to either fix or write equivalent checks.
+	// fso->size may be irrelevant now. Or not?
+	//assert(fso->size <= fso->allocated);
+	//assert(t->offset <= fso->size);
+
+	unsigned int total_offset = 0;
+	struct file_data *current = (struct file_data*) fso->data;
+	
+	/* file does not exist yet */
+	if (current == NULL) {
+		struct file_data *new_node;
+		file_data_alloc(&new_node, cb, 0, sz, start, sz, NULL);
+
+		fso->data = new_node;
+
+		t->offset += sz;
+		fso->size = t->offset;
+
+		ret = sz;
+
+		goto done;
+	}
+	
+	unsigned int curr_off = 0;
+	struct file_data *prev = NULL;
+
+	/* iterate through and try to find where to write to */
+	while (current != NULL &&
+		(total_offset + current->len) <= t->offset) {
+		total_offset += current->len;
+
+		prev = current;
+		current = current->next;
+	}
+
+	curr_off = t->offset - total_offset;
+
+	if (current == NULL) {
+		// arguably the most likely case
+		// just writing to the end of the file
+		struct file_data *new_node;
+		file_data_alloc(&new_node, cb, 0, sz, start, sz, NULL);
+
+		prev->next = new_node;
+
+		t->offset += sz;
+		fso->size = t->offset;
+	}
+	else if (sz < current->len - curr_off) {
+		unsigned int offset_into_cbuf = current->start + curr_off;
+		int split_len = current->len - offset_into_cbuf - sz;
+		int copy_start = current->start + offset_into_cbuf + sz;
+		cbuf_t split_cb;
+		char *cb_ptr = cbuf_alloc(split_len, &split_cb);
+		if (!cb_ptr) { ERR_THROW(-EINVAL, done); }
+		char *buf = cbuf2buf(current->cbuf_id, current->len);
+		memcpy(cb_ptr, buf + copy_start, split_len);
+
+		// create a node for this new cbuf.
+		struct file_data *split_node;
+		file_data_alloc(&split_node, split_cb, 0, split_len, 0, split_len, current->next);
+
+		// now create a node for the cbuf they passed in.
+		struct file_data *new_node;
+		file_data_alloc(&new_node, cb, 0, sz, start, sz, split_node);
+
+		// modify current
+		current->len -= (sz + split_len);
+		current->next = new_node;
+	}
+	else if (sz > current->len) {
+		struct file_data* ptr_node = current->next;
+		struct file_data* head_node = current;
+		if (curr_off == 0)
+		{
+			ptr_node = current;
+			head_node = prev;
+		}
+
+		int bytesOverwritten = current->len - curr_off;
+		current->len -= bytesOverwritten;
+
+		while (ptr_node != NULL && bytesOverwritten + ptr_node->len
+			< sz) {
+			bytesOverwritten += ptr_node->len;
+			struct file_data *temp_ptr = ptr_node;
+			ptr_node = ptr_node->next;
+			free(temp_ptr);
+		}
+
+		if (bytesOverwritten < sz && ptr_node != NULL) {
+			ptr_node->start += (sz - bytesOverwritten);
+		}
+
+		struct file_data *new_node;
+		file_data_alloc(&new_node, cb, 0, sz, start, sz, ptr_node);
+
+		head_node->next = new_node;
+	}
+	else {
+		printc(">>>case 5: did not match anything!!!\n");
+		printc("t->offset: %d\n", t->offset);
+		ERR_THROW(-EINVAL, done);
+	}
+
+	// It should never equal anything else?
+	ret = sz;
+
```

Comment:
my feeling is that all of this needs a thorough revisit, and a thorough test case to make sure that all cases work.  5 cases seems like an unfortunately large amount.  perhaps the techniques we discussed before could simplify.